### PR TITLE
Capture Click outside Mega Menu

### PR DESCRIPTION
**Pull request by debsmita1**
This PR contains:
The Nav Toggle button is highlighted when hovered or Perspective Switcher is Open.
The Perspective Switcher closes when clicked outside.

Nav-toggle

Refer task : https://jira.coreos.com/browse/ODC-237

### CODE REVIEW

File: frontend/public/extend/devconsole/shared/components/MegaMenu/MegaMenu.tsx [hunk 1]
```diff
@@ -6,13 +6,48 @@ import './MegaMenu.scss';
...
-  </React.Fragment>
-);
+export default class MegaMenu extends React.Component<MegaMenuProps> {
+  constructor(props: MegaMenuProps) {
+    super(props);
+    this._handleClickOutside = this._handleClickOutside.bind(this);
+  }
 
-export default MegaMenu;
+  public node: any;
```
Comment (rohitkrai03) [inline frontend/public/extend/devconsole/shared/components/MegaMenu/MegaMenu.tsx]: Add type for `node`

File: frontend/public/style/_overrides.scss [hunk 2]
```diff
@@ -301,6 +301,21 @@ tags-input .autocomplete .suggestion-item em {
   }
 }
 
+.nav-toggle {
```
Comment (rohitkrai03) [inline frontend/public/style/_overrides.scss]: Don't add the styles in overrides.scss file. Better to add this to `PerspectiveSwitcher.scss` or `MegaMenu.scss`

File: frontend/public/style/_overrides.scss [hunk 3]
```diff
@@ -301,6 +301,21 @@ tags-input .autocomplete .suggestion-item em {
...
+  &__active {
+    .pf-c-button {
+      background-color: grey !important ;
```
Comment (rohitkrai03) [inline frontend/public/style/_overrides.scss]: Use patternfly 4 css variables for color

File: frontend/public/style/_overrides.scss [hunk 4]
```diff
@@ -301,6 +301,21 @@ tags-input .autocomplete .suggestion-item em {
   }
 }
 
+.nav-toggle {
+  &__active {
+    .pf-c-button {
```
Comment (divyanshiGupta) [inline frontend/public/style/_overrides.scss]: The document https://docs.google.com/document/d/1n-Sg7trE9x3cD4jUZNfZmGLKm2y1D7280aO8zWXw8q8/edit?pli=1#heading=h.rz8hwgg5vy3k shows the color to be white. And we should use patternfly variables where ever possible. You can find the variables here https://patternfly-react.surge.sh/patternfly-4/styles/tokens, just add `background-color: var(variable-name)` for example or refer my PR if you have any confusions.

File: frontend/public/style/_overrides.scss [hunk 5]
```diff
@@ -301,6 +301,21 @@ tags-input .autocomplete .suggestion-item em {
...
+      background-color: grey !important ;
+      border-radius: 25px !important ;
+    }
+  }
+  &__hover {
+    .pf-c-button:hover {
+      background-color: grey ;
```
Comment (divyanshiGupta) [inline frontend/public/style/_overrides.scss]: Same as above

File: frontend/public/style/_overrides.scss [hunk 6]
```diff
@@ -301,6 +301,21 @@ tags-input .autocomplete .suggestion-item em {
...
+      border-radius: 25px ;
```
Comment (divyanshiGupta) [inline frontend/public/style/_overrides.scss]: See if you can use variables for `border-radius` as well.

File: frontend/public/components/masthead.jsx [hunk 7]
```diff
@@ -48,29 +48,42 @@ export const getBrandingDetails = () => {
...
 
-export const Masthead = ({ defaultRoute, onNavToggle }) => {
+export const Masthead = ({
+  isPerspectiveSwitcherActive,
+  defaultRoute,
+  isNavOpen,
+  onNavToggle
+}) => {
   const details = getBrandingDetails();
+  const defaultRoute = "/";
```
[the remark this answers is not in the document]
Comment (debsmita1) [reply frontend/public/components/masthead.jsx]: I overlooked this one while rebasing. Will remove this.

File: frontend/public/style/_overrides.scss (hunk 2 above)
Comment (debsmita1) [reply frontend/public/style/_overrides.scss]: PerspectiveSwitcher.scss file has been removed. So will add styles in MegaMenu.scss

File: frontend/public/extend/devconsole/shared/components/MegaMenu/MegaMenu.tsx [hunk 8]
```diff
@@ -6,13 +6,48 @@ import './MegaMenu.scss';
...
+  public navToggleButton: HTMLElement;
+
+  public _handleClickOutside = (event: MouseEvent) => {
```
[the remark this answers is not in the document]
Comment (debsmita1) [reply frontend/public/extend/devconsole/shared/components/MegaMenu/MegaMenu.tsx]: Done

File: frontend/public/extend/devconsole/shared/components/MegaMenu/MegaMenu.tsx (hunk 1 above)
Comment (debsmita1) [reply frontend/public/extend/devconsole/shared/components/MegaMenu/MegaMenu.tsx]: Done

File: frontend/public/style/_overrides.scss (hunk 2 above)
Comment (debsmita1) [reply frontend/public/style/_overrides.scss]: Done

File: frontend/public/style/_overrides.scss (hunk 3 above)
Comment (debsmita1) [reply frontend/public/style/_overrides.scss]: Done

File: frontend/public/extend/devconsole/shared/components/MegaMenu/MegaMenu.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (debsmita1) [reply frontend/public/extend/devconsole/shared/components/MegaMenu/MegaMenu.tsx]: Done

File: frontend/public/components/masthead.jsx [hunk 9]
```diff
@@ -48,29 +48,41 @@ export const getBrandingDetails = () => {
...
     <PageHeader
+      className={[
+        isPerspectiveSwitcherActive ? 'nav-toggle__hover' : '',
+        isPerspectiveSwitcherActive && isNavOpen ? 'nav-toggle__active' : ''
+      ]}
       logo={<Brand src={details.logoImg} alt={details.logoAlt} />}
       logoProps={logoProps}
       toolbar={<MastheadToolbar />}
       showNavToggle
+      isNavOpen={isNavOpen}
```
Comment (divyanshiGupta) [inline frontend/public/components/masthead.jsx]: Why do we need to pass this prop to header?

File: frontend/public/extend/devconsole/shared/components/MegaMenu/MegaMenu.scss [hunk 10]
```diff
@@ -19,3 +19,18 @@
 .backdrop {
   position: absolute !important;
 }
+
+.nav-toggle {
```
Comment (divyanshiGupta) [inline frontend/public/extend/devconsole/shared/components/MegaMenu/MegaMenu.scss]: change your classname to `odc-mega-menu-toggle`

**Comment by divyanshiGupta** [comment]
Run linter and fix errors
Use prettier to format your files

File: frontend/public/extend/devconsole/shared/components/MegaMenu/MegaMenu.scss [hunk 11]
```diff
@@ -19,3 +19,18 @@
...
+  &__active {
+    .pf-c-button {
+      background-color: var(--pf-global--disabled-color--100);
+      border-radius: 25px;
+    }
+  }
+  &__hover {
```
Comment (divyanshiGupta) [inline frontend/public/extend/devconsole/shared/components/MegaMenu/MegaMenu.scss]: Why do you need two make two different classes with same styles? You can use one class of these classes and add it when you hover over the button and when the mega-menu is open

File: frontend/public/style/_overrides.scss (hunk 4 above)
Comment (debsmita1) [reply frontend/public/style/_overrides.scss]: Done

File: frontend/public/style/_overrides.scss (hunk 5 above)
Comment (debsmita1) [reply frontend/public/style/_overrides.scss]: Done

File: frontend/public/style/_overrides.scss (hunk 6 above)
Comment (debsmita1) [reply frontend/public/style/_overrides.scss]: The patternfly's border radius variable has value 3px. So kept my code as it is.

File: frontend/public/components/masthead.jsx (hunk 9 above)
Comment (debsmita1) [reply frontend/public/components/masthead.jsx]: Removed it.

File: frontend/public/extend/devconsole/shared/components/MegaMenu/MegaMenu.scss (hunk 10 above)
Comment (debsmita1) [reply frontend/public/extend/devconsole/shared/components/MegaMenu/MegaMenu.scss]: Done.

**Comment by divyanshiGupta** [comment]
I pulled your PR and tried running it. I cannot see the button getting highlighted.

**Comment by divyanshiGupta** [comment]
Can you write some unit tests for the outside click functionality?

File: frontend/public/extend/devconsole/shared/components/MegaMenu/MegaMenu.tsx [hunk 12]
```diff
@@ -6,13 +6,47 @@ import './MegaMenu.scss';
...
+  node: HTMLElement;
+
+  _handleClickOutside = (event: MouseEvent) => {
+    if (
+      this.props.isNavOpen &&
+      this.node &&
+      !this.node.contains(event.target as HTMLElement) &&
+      !(event.target as HTMLElement).className.includes('pf-c-button')
+    ) {
+      this.props.onNavToggle(event);
```
Comment (christianvogt) [inline frontend/public/extend/devconsole/shared/components/MegaMenu/MegaMenu.tsx]: Use `this.props.onClose()`

File: frontend/public/extend/devconsole/shared/components/MegaMenu/MegaMenu.tsx [hunk 13]
```diff
@@ -6,13 +6,47 @@ import './MegaMenu.scss';
...
-const MegaMenu: React.FunctionComponent<MegaMenuProps> = (props: MegaMenuProps) => (
-  <React.Fragment>
-    {props.isNavOpen && <Backdrop className="backdrop" />}
-    <div className={`odc-mega-menu ${props.isNavOpen ? 'is-open' : ''}`}>{props.children}</div>
-  </React.Fragment>
-);
+export default class MegaMenu extends React.Component<MegaMenuProps> {
+  constructor(props: MegaMenuProps) {
+    super(props);
+    this._handleClickOutside = this._handleClickOutside.bind(this);
```
Comment (christianvogt) [inline frontend/public/extend/devconsole/shared/components/MegaMenu/MegaMenu.tsx]: Unnecessary to do this binding in typescript.
You can remove this constructor.

File: frontend/public/extend/devconsole/shared/components/MegaMenu/MegaMenu.tsx [hunk 14]
```diff
@@ -6,13 +6,47 @@ import './MegaMenu.scss';
...
+  }
 
-export default MegaMenu;
+  node: HTMLElement;
+
+  _handleClickOutside = (event: MouseEvent) => {
```
Comment (christianvogt) [inline frontend/public/extend/devconsole/shared/components/MegaMenu/MegaMenu.tsx]: We can use `private` functions in typescript.

File: frontend/public/extend/devconsole/shared/components/MegaMenu/MegaMenu.tsx [hunk 15]
```diff
@@ -6,13 +6,47 @@ import './MegaMenu.scss';
...
-  </React.Fragment>
-);
+export default class MegaMenu extends React.Component<MegaMenuProps> {
+  constructor(props: MegaMenuProps) {
+    super(props);
+    this._handleClickOutside = this._handleClickOutside.bind(this);
+  }
 
-export default MegaMenu;
+  node: HTMLElement;
```
Comment (christianvogt) [inline frontend/public/extend/devconsole/shared/components/MegaMenu/MegaMenu.tsx]: We can use `private` properties in typescript.

File: frontend/public/extend/devconsole/shared/components/MegaMenu/MegaMenu.tsx [hunk 16]
```diff
@@ -6,13 +6,47 @@ import './MegaMenu.scss';
...
+      this.node &&
+      !this.node.contains(event.target as HTMLElement) &&
+      !(event.target as HTMLElement).className.includes('pf-c-button')
+    ) {
+      this.props.onNavToggle(event);
+    }
+  };
+
+  componentWillMount() {
+    document.addEventListener('mousedown', this._handleClickOutside);
```
Comment (christianvogt) [inline frontend/public/extend/devconsole/shared/components/MegaMenu/MegaMenu.tsx]: We should probably only attach the listener when the menu is open. Also use the `click` event.
```
  componentDidUpdate() {
    if (this.props.isNavOpen) {
      document.addEventListener('click', this.handleClickOutside);
    } else {
      document.removeEventListener('click', this.handleClickOutside);
    }
  }
```

File: frontend/public/components/masthead.jsx [hunk 17]
```diff
@@ -1,66 +1,77 @@
...
   };
 
   return (
     <PageHeader
+      className={[
+        isPerspectiveSwitcherActive ? "odc-mega-menu-toggle__hover" : "",
+        isPerspectiveSwitcherActive && isNavOpen
+          ? "odc-mega-menu-toggle__active"
+          : ""
+      ]}
```
Comment (christianvogt) [inline frontend/public/components/masthead.jsx]: Do not use arrays in `className`.
```
import cx from 'classnames';
```
```
className={cx({
  'odc-mega-menu-toggle': isPerspectiveSwitcherActive,
  'is-active': isPerspectiveSwitcherActive && isNavOpen
})}
```

File: frontend/public/extend/devconsole/shared/components/MegaMenu/MegaMenu.scss [hunk 18]
```diff
@@ -19,3 +19,20 @@
...
+    .pf-c-button:hover {
+      background-color: var(--pf-global--BackgroundColor--light-100);
+      border-radius: 25px;
+      color: var(--pf-global--BackgroundColor--dark-100);
+    }
+  }
+}
```
Comment (christianvogt) [inline frontend/public/extend/devconsole/shared/components/MegaMenu/MegaMenu.scss]: `pf-c-button` is too generic. Since this is just a temporary fix, you can reference the `#nav-toggle` element ID.

```
.odc-mega-menu-toggle {
  &.is-active #nav-toggle {
      background-color: var(--pf-global--BackgroundColor--light-100);
      border-radius: 25px;
      color: var(--pf-global--BackgroundColor--dark-100);
    }
    & #nav-toggle:hover {
      background-color: var(--pf-global--BackgroundColor--light-100);
      border-radius: 25px;
      color: var(--pf-global--BackgroundColor--dark-100);
    }
}
```

File: frontend/public/extend/devconsole/shared/components/MegaMenu/MegaMenu.tsx [hunk 19]
```diff
@@ -6,13 +6,47 @@ import './MegaMenu.scss';
...
+  }
+
+  componentWillUnmount() {
+    document.removeEventListener('mousedown', this._handleClickOutside);
```
Comment (christianvogt) [inline frontend/public/extend/devconsole/shared/components/MegaMenu/MegaMenu.tsx]: Use `click` event.

File: frontend/public/components/app.jsx [hunk 20]
```diff
@@ -203,7 +203,7 @@ class App extends React.PureComponent {
           defaultTitle={productName}
         />
         <ConsolePage
-          header={<Masthead defaultRoute={defaultRoute} onNavToggle={this._onNavToggle} />}
+          header={<Masthead isPerspectiveSwitcherActive={this.props.flags.SHOW_DEV_CONSOLE} defaultRoute={defaultRoute} isNavOpen={!isNavOpen} onNavToggle={this._onNavToggle} />}
```
[the remark this answers is not in the document]
Comment (debsmita1) [reply frontend/public/components/app.jsx]: @christianvogt  Sorry, but I didn't get this. I understand why we need `onClose` as a prop in the MegaMenu component, as the menu is notified to close and never to open when clicked outside the Perspective Switcher. But in the MastHead and Perspective Switcher Component we need the `onNavtoggle` prop.

And regarding this `isNavOpen: this.props.flags.SHOW_DEV_CONSOLE ? false : this._isDesktop(),`
we cannot initialize `isNavProp` in the constructor, with the above, because `this.props.flags.SHOW_DEV_CONSOLE` is undefined since devconsole Operator is detected much later as a result Perspective Switcher is active everytime the Application is loaded

File: frontend/public/extend/devconsole/shared/components/MegaMenu/MegaMenu.tsx [hunk 21]
```diff
@@ -6,13 +6,47 @@ import './MegaMenu.scss';
...
 
-export default MegaMenu;
+  node: HTMLElement;
+
+  _handleClickOutside = (event: MouseEvent) => {
+    if (
+      this.props.isNavOpen &&
+      this.node &&
+      !this.node.contains(event.target as HTMLElement) &&
+      !(event.target as HTMLElement).className.includes('pf-c-button')
```
[the remark this answers is not in the document]
Comment (debsmita1) [reply frontend/public/extend/devconsole/shared/components/MegaMenu/MegaMenu.tsx]: @christianvogt  I did this:

`private _handleClickOutside = (event: MouseEvent) => {`
   ` if (this.node && !this.node.contains(event.target as HTMLElement)) {`
      `this.props.onClose();`
   ` }`
  `};`
But `isNavOpen` 's state doesnt seem to change to false when clicked outside.

File: frontend/public/components/app.jsx (hunk 20 above)
[the remark this answers is not in the document]
Comment (christianvogt) [reply frontend/public/components/app.jsx]: When does the `PerspectiveSwitcher` need to toggle itself open? The `Masthead` needs it because it owns the hamburger button.

Oh right the `SHOW_DEV_CONSOLE` flag is lazy initialized. Argh. So we cannot use this in the constructor. It would be best then to separate out `isNavOpen` into a separate states for the regular nav and another for the perspective switcher. Add another state called `isPerspectiveSwitcherOpen` initialized to `false` in the constructor. (I just tried this out)

```
  _onNavToggle() {
    this.setState(prevState => {
      return {
        isNavOpen: this.props.flags.SHOW_DEV_CONSOLE
          ? prevState.isNavOpen
          : !prevState.isNavOpen,
        isPerspectiveSwitcherOpen: this.props.flags.SHOW_DEV_CONSOLE
          ? !prevState.isPerspectiveSwitcherOpen
          : false,
      };
    });
  }

  _onPerspectiveSwitcherClose() {
    this.setState({ isPerspectiveSwitcherOpen: false });
  }
```

We can now remove the checks for `SHOW_DEV_CONSOLE` from these functions:
```
  _onNavSelect() {
    //close nav on mobile nav selects
    if (!this._isDesktop()) {
      this.setState({ isNavOpen: false });
    }
  }

  _onResize() {
    const isDesktop = this._isDesktop();
    if (this.previousDesktopState !== isDesktop) {
      this.setState({ isNavOpen: isDesktop });
      this.previousDesktopState = isDesktop;
    }
  }
```

We need to update how the sidebar works:
```
  _sidebarNav() {
    if (
      this.props.flags.SHOW_DEV_CONSOLE &&
      this.props.activePerspective === 'dev'
    ) {
      return <DevConsoleNavigation isNavOpen={true} />;
    }
    return (
      <Navigation
        isNavOpen={this.state.isNavOpen}
        onNavSelect={this._onNavSelect}
      />
    );
  }
```

And finally we need to update the `app.jsx#render` function to use `isPerspectiveSwitcherOpen`:
```
        <ConsolePage
          header={
            <Masthead
              isPerspectiveSwitcherActive={this.props.flags.SHOW_DEV_CONSOLE}
              defaultRoute={defaultRoute}
              isNavOpen={isPerspectiveSwitcherOpen}
              onNavToggle={this._onNavToggle}
            />
          }
          sidebar={this._sidebarNav()}
          megaMenu={
            <PerspectiveSwitcher
              isNavOpen={isPerspectiveSwitcherOpen}
              onClose={this._onPerspectiveSwitcherClose}
            />
          }
        >
```

**Comment by divyanshiGupta** [comment]
@debsmita1 you are supposed to use prettier for our code only. You have used prettier on some of the console files like `masthead.jsx`.

File: frontend/public/extend/devconsole/shared/components/MegaMenu/MegaMenu.tsx [hunk 22]
```diff
@@ -6,13 +6,47 @@ import './MegaMenu.scss';
...
+  onNavToggle: (MouseEvent) => void;
 }
 
-const MegaMenu: React.FunctionComponent<MegaMenuProps> = (props: MegaMenuProps) => (
-  <React.Fragment>
-    {props.isNavOpen && <Backdrop className="backdrop" />}
-    <div className={`odc-mega-menu ${props.isNavOpen ? 'is-open' : ''}`}>{props.children}</div>
-  </React.Fragment>
-);
+export default class MegaMenu extends React.Component<MegaMenuProps> {
```
Comment (christianvogt) [inline frontend/public/extend/devconsole/shared/components/MegaMenu/MegaMenu.tsx]: Extend `React.PureComponent` to prevent needless re-renders from parent components.

File: frontend/public/components/masthead.jsx [hunk 23]
```diff
@@ -9,58 +9,68 @@ import openshiftLogoImg from '../imgs/openshift-logo.svg';
...
       logoAlt = 'OpenShift Dedicated';
       productTitle = 'Red Hat OpenShift Dedicated';
       break;
-    case 'azure':
-      logoImg = azureLogoImg;
-      logoAlt = 'Azure Red Hat OpenShift';
-      productTitle = 'Azure Red Hat OpenShift';
-      break;
     default:
+      backgroundImg = pfBg992;
```
Comment (christianvogt) [inline frontend/public/components/masthead.jsx]: Why are these changes here. This shouldn't be part of your pr.

File: frontend/public/extend/devconsole/shared/components/MegaMenu/MegaMenu.tsx [hunk 24]
```diff
@@ -6,13 +6,37 @@ import './MegaMenu.scss';
...
+    }
+  };
+
+  componentDidUpdate() {
+    if (this.props.isNavOpen) {
+      document.addEventListener('click', this._handleClickOutside);
+    } else {
+      document.removeEventListener('click', this._handleClickOutside);
+    }
+  }
```
Comment (christianvogt) [inline frontend/public/extend/devconsole/shared/components/MegaMenu/MegaMenu.tsx]: You still need to remove your listener when unmounting.

File: frontend/public/extend/devconsole/shared/components/__tests__/MegaMenu.spec.tsx [hunk 25]
```diff
@@ -0,0 +1,25 @@
+/* eslint-disable no-unused-vars, no-undef */
+import * as React from 'react';
+import { shallow, ShallowWrapper } from 'enzyme';
+import MegaMenu from '../MegaMenu/MegaMenu';
+import { Backdrop } from '@patternfly/react-core';
+
```
Comment (christianvogt) [inline frontend/public/extend/devconsole/shared/components/__tests__/MegaMenu.spec.tsx]: Isn't this class part of Divyanshi's pr already? Why would it show up.

Need some tests for the click outside functionality.

Comment (debsmita1) [reply frontend/public/extend/devconsole/shared/components/__tests__/MegaMenu.spec.tsx]: Rebasing my code.

File: frontend/public/components/masthead.jsx (hunk 23 above)
Comment (debsmita1) [reply frontend/public/components/masthead.jsx]: I had mistakenly applied prettier on this file before. This  happened while undoing the changes made by prettier. fixing it

File: frontend/public/extend/devconsole/shared/components/__tests__/MegaMenu.spec.tsx (hunk 25 above)
Comment (debsmita1) [reply frontend/public/extend/devconsole/shared/components/__tests__/MegaMenu.spec.tsx]: Yes I am writing tests for outside click functionality

File: frontend/public/extend/devconsole/shared/components/ConsolePage.scss [hunk 26]
```diff
@@ -1,4 +1,4 @@
-.openshift-console-page {
+.odc-console-page {
```
Comment (divyanshiGupta) [inline frontend/public/extend/devconsole/shared/components/ConsolePage.scss]: Why is this change showing up here? Its already in my PR. I think there is some issue with rebasing

File: frontend/public/extend/devconsole/shared/components/ConsolePage.tsx [hunk 27]
```diff
@@ -14,9 +14,9 @@ const ConsolePage: React.FunctionComponent<ConsolePageProps> = (props: ConsolePa
     <div className="pf-c-page">
       {props.header}
       <main role="main" className="pf-c-page__main">
-        <div className="openshift-console-page">
+        <div className="odc-console-page">
           <div style={{ display: 'flex' }}>{props.sidebar}</div>
-          <div className="openshift-console-page__section">{props.children}</div>
+          <div className="odc-console-page__section">{props.children}</div>
```
Comment (divyanshiGupta) [inline frontend/public/extend/devconsole/shared/components/ConsolePage.tsx]: Rebasing issue

File: frontend/public/extend/devconsole/shared/components/__tests__/MegaMenuItem.spec.tsx [hunk 28]
```diff
@@ -0,0 +1,26 @@
+/* eslint-disable no-unused-vars, no-undef */
+import * as React from 'react';
+import { shallow, ShallowWrapper } from 'enzyme';
+import MegaMenuItem from '../MegaMenu/MegaMenuItem';
+import { ExternalLinkAltIcon } from '@patternfly/react-icons';
+
```
Comment (divyanshiGupta) [inline frontend/public/extend/devconsole/shared/components/__tests__/MegaMenuItem.spec.tsx]: Rebasing issue. This should not appear in this PR

File: frontend/public/extend/devconsole/shared/components/__tests__/PerspectiveSwitcher.spec.tsx [hunk 29]
```diff
@@ -1,54 +1,40 @@
...
-      <PerspectiveSwitcher
-        isNavOpen={true}
-        activePerspective="admin"
-        onNavToggle={() => {}}
-        flags={{ SHOW_DEV_CONSOLE: true }}
-        onPerspectiveChange={() => {}}
-      />,
-    );
-    expect(switcherWrapper.find(Modal).prop('isOpen')).toEqual(true);
+    expect(switcherWrapper.find(MegaMenu).length).toEqual(1);
```
Comment (divyanshiGupta) [inline frontend/public/extend/devconsole/shared/components/__tests__/PerspectiveSwitcher.spec.tsx]: Rebasing issue

File: frontend/public/extend/devconsole/shared/components/__tests__/PerspectiveSwitcher.spec.tsx [hunk 30]
```diff
@@ -1,54 +1,40 @@
...
 import { UIActions } from '../../../../../ui/ui-actions';
 import ConnectedPerspectiveSwitcher, { PerspectiveSwitcher } from '../PerspectiveSwitcher';
 import { getStoreTypedComponent } from '../../../utils/test-utils';
+import MegaMenu from '../MegaMenu/MegaMenu';
 
 describe('PerspectiveSwitcher', () => {
   let switcherWrapper: ShallowWrapper<any> = null;
 
-  it('renders perspective switcher menu and it should be closed initially', () => {
+  it('renders perspective switcher menu', () => {
```
Comment (divyanshiGupta) [inline frontend/public/extend/devconsole/shared/components/__tests__/PerspectiveSwitcher.spec.tsx]: Rebasing issue

**Comment by divyanshiGupta** [comment]
@debsmita1 you need to properly rebase your branch against mine and resolve conflicts very carefully. Because of some rebasing issue you were not able to see few of my changes. When you see conflicts in some file, refer my PR before making any changes.

**Comment by divyanshiGupta** [comment]
@debsmita1 @christianvogt I am getting this error while running the latest code from this PR - 
```
ERROR in /home/divgupta/go/src/github.com/openshift/console/frontend/public/extend/devconsole/shared/components/__tests__/MegaMenu.spec.tsx(58,21):
TS2304: Cannot find name 'jest'.

```
I tried adding `@types/jest` but got a bunch of new errors related to jest and jasmine `.d.ts` files.

**Comment by divyanshiGupta** [comment]
@christianvogt @debsmita1 Never mind. Just solved it by adding `@types/jest` and `"skipLibCheck": true` in `tsconfig.json`. Found this while going through `tsconfig.json` of OSIO. ;-)

**Comment by christianvogt** [comment]
@divyanshiGupta You're correct that the type should be present in `tsconfig` and should also be a dev dependency in the `package.json`.
